### PR TITLE
pybind/mgr: fix "osd status" command output with format json

### PR DESCRIPTION
**Pull request by NitzanMordhai**
currently command "ceph -f json osd status" return table view instead of json view.
the fix will fix that and will also fix the json-pretty format output.

Signed-off-by: Nitzan Mordechai <nmordech@redhat.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [x] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by NitzanMordhai** [comment]
> 

jenkins test make check

**Review by batrick — CHANGES_REQUESTED** [review]
Please use `pybind/mgr` or `pybind/mgr/status`.

**Review by joscollin — CHANGES_REQUESTED** [review]
As you are on it, fix the `xml` and xml-pretty formats too?

**Comment by NitzanMordhai** [comment]
> As you are on it, fix the `xml` and xml-pretty formats too?

Does it work for any other command output that came from python modules?
anyway, I think we need new tracker for that one

**Comment by joscollin** [comment]
> > As you are on it, fix the `xml` and xml-pretty formats too?
> 
> Does it work for any other command output that came from python modules? 

Yes, check out other commands.

> anyway, I think we need new tracker for that one

It doesn't need a tracker, as there is no need for a backport afaik.

**Comment by joscollin** [comment]
> > As you are on it, fix the `xml` and xml-pretty formats too?
> 
> Does it work for any other command output that came from python modules? anyway, I think we need new tracker for that one


Making it similar to `fs status` options is fine.

**Comment by NitzanMordhai** [comment]
> > > As you are on it, fix the `xml` and xml-pretty formats too?
> > 
> > 
> > Does it work for any other command output that came from python modules?
> 
> Yes, check out other commands.
> 
All the python handled command doesn't have xml\xml-pretty output formatter, the command you tested coming from c++ code.
we will leave the python code for now without xml\xml-pretty formats and if we want to add it later we will need to think of class that will do it like the c++ approach 

> > anyway, I think we need new tracker for that one
> 
> It doesn't need a tracker, as there is no need for a backport afaik.

**Review by joscollin — APPROVED** [review]

File: src/pybind/mgr/status/module.py [hunk 1]
```diff
@@ -349,5 +353,22 @@ def handle_osd_status(self, bucket: Optional[str] = None) -> Tuple[int, str, str
                                mgr_util.format_bytes(rd_byte_rate, 5),
                                ','.join(osd['state']),
                                ])
-
-        return 0, osd_table.get_string(), ""
+            if output_format in ('json', 'json-pretty'):
```
[the remark this answers is not in the document]
Comment (NitzanMordhai) [reply src/pybind/mgr/status/module.py]: Done

File: src/pybind/mgr/status/module.py [hunk 2]
```diff
@@ -349,5 +353,22 @@ def handle_osd_status(self, bucket: Optional[str] = None) -> Tuple[int, str, str
...
+                                        'host name': hostname,
+                                        'kb used' : kb_used,
+                                        'kb available':kb_avail,
+                                        'write ops rate': wr_ops_rate,
+                                        'write byte rate': wr_byte_rate,
+                                        'read ops rate': rd_ops_rate,
+                                        'read byte rate': rd_byte_rate,
+                                        'state': osd['state']
+                                      })
+
```
[the remark this answers is not in the document]
Comment (NitzanMordhai) [reply src/pybind/mgr/status/module.py]: Done

File: src/pybind/mgr/status/module.py [hunk 3]
```diff
@@ -279,10 +279,14 @@ def handle_fs_status(self,
             return HandleCommandResult(stdout=output)
 
     @CLIReadCommand("osd status")
-    def handle_osd_status(self, bucket: Optional[str] = None) -> Tuple[int, str, str]:
+    def handle_osd_status(self, bucket: Optional[str] = None, format: str = 'plain') -> Tuple[int, str, str]:
         """
         Show the status of OSDs within a bucket, or all
         """
+        json_output: Dict[str, List[str]] = \
```
[the remark this answers is not in the document]
Comment (NitzanMordhai) [reply src/pybind/mgr/status/module.py]: Yes, changed that, thanks

File: src/pybind/mgr/status/module.py [hunk 4]
```diff
@@ -279,10 +279,14 @@ def handle_fs_status(self,
             return HandleCommandResult(stdout=output)
 
     @CLIReadCommand("osd status")
-    def handle_osd_status(self, bucket: Optional[str] = None) -> Tuple[int, str, str]:
+    def handle_osd_status(self, bucket: Optional[str] = None, format: str = 'plain') -> Tuple[int, str, str]:
         """
         Show the status of OSDs within a bucket, or all
         """
+        json_output: Dict[str, List[Any]] = \
```
Comment (kotreshhr) [inline src/pybind/mgr/status/module.py:286]: need to import `Any`

**Comment by NitzanMordhai** [comment]
@batrick can you please review the changes?

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
@batrick hi Patrick, i believe your comment has been addressed. could you take another look?

**Comment by tchaikov** [comment]
jenkins test make check

**Review by batrick — APPROVED** [review]

**Pull request merged**